tests/odm/test_encoder.py

```diff
@@ -125,3 +128,30 @@ def test_keep_nulls_false():
     encoder = Encoder(keep_nulls=False, to_db=True)
     encoded_doc = encoder.encode(doc)
     assert encoded_doc == {"m": {"i": 10}}
+
+
+async def test_url():
+    # same logic from encoder.py IS_PYDANTIC_V2
+    is_pydantic_v2 = int(pydantic.VERSION.split(".")[0]) >= 2
+    if is_pydantic_v2:
+        doc = DocumentWithHttpUrlField(url_field="https://example.com")
+        assert isinstance(doc.url_field, Url)
+        await doc.save()
+
+        new_doc = await DocumentWithHttpUrlField.find_one(
+            DocumentWithHttpUrlField.id == doc.id
+        )
+
+        assert isinstance(new_doc.url_field, Url)
+        assert new_doc.url_field == doc.url_field
+    else:
+        doc = DocumentWithHttpUrlField(url_field="https://example.com")
+        assert isinstance(doc.url_field, str)
```

Comment:
Please check, if Pydantic V1 has a class for URL inherited from str. I'll check it too

Reply:
@roman-right 

Yes it has `AnyUrl` class.
Is it better to assert `AnyUrl` instead of `str` ?

```python
def test_debug():

	model = HttpUrlModel(url="https://www.google.com")
	assert model.url == "https://www.google.com"
	assert isinstance(model.url, AnyUrl) # True since HttpUrl(AnyHttpUrl(AnyUrl(str)))
```

Reply:
Yes, it is better, as it will display, that it was deserialized.
Hm. Will not `HttpUrl` work for both cases here then?

Reply:
Or maybe I'm missing the point of the test

Reply:
> Yes, it is better, as it will display, that it was deserialized.
> Hm. Will not HttpUrl work for both cases here then?

No it won't work since v1 use class inheritance but v2 uses subscripted generics for `HttpUrl`.

v2 networks.py
```python
AnyUrl = Url
"""A type that will accept any http or https URL."""
HttpUrl = Annotated[Url, UrlConstraints(max_length=2083, allowed_schemes=['http', 'https'])]
```
v2 _pydantic_core.py
```python
class Url(object):
```
v1 networks.py
```python
class AnyUrl(str):
```


But if we use `AnyUrl` as assertion it will pass both version without if condition though it hides the base class of instance.
```python
async def test_url():
    doc = DocumentWithHttpUrlField(url_field="https://example.com")
    assert isinstance(doc.url_field, AnyUrl)
    await doc.save()

    new_doc = await DocumentWithHttpUrlField.find_one(
        DocumentWithHttpUrlField.id == doc.id
    )

    assert isinstance(new_doc.url_field, AnyUrl)
    assert new_doc.url_field == doc.url_field
```

I think this depends on beanie's policy what we should assert exactly.

Reply:
I think we have to be sure that it was encoded correctly. How it was decoded doesn't matter that much.
I'd check how the encoder encodes the document and would check if the output is correct. This way, we will be sure that it was stored in the db correctly. How it will be parsed is more Pydantic's responsibility. What do you think?

Reply:
Yes, I agree.
I've added a test for encoding.
Saving/Retrieving test case is remained to confirm it works in practice.